### ISSUE TITLE
Tcl vs. Python Comparative Case Study

### ISSUE THREAD

**Issue by hs-apotell** [chipsalliance/Surelog#2214]
Python implementation has been running in parallel to the Tcl implementation for sometime now and so it's a good time to do a comparative case study of both implementation. I have used the following builds to collect all the data below. Note, that builds in each row are being run on exactly the same code base. Also, note that both Surelog and UHDM are using Python implementation and/or Tcl implementation respectively i.e. no mix and match within the same build.

* Tcl [2614](https://github.com/chipsalliance/Surelog/actions/runs/1458231570) vs. Py [94](https://github.com/chipsalliance/Surelog/actions/runs/1458231572)
* Tcl [2610](https://github.com/chipsalliance/Surelog/actions/runs/1457435043) vs. Py [90](https://github.com/chipsalliance/Surelog/actions/runs/1457435044)
* Tcl [2606](https://github.com/chipsalliance/Surelog/actions/runs/1456020783) vs. Py [86](https://github.com/chipsalliance/Surelog/actions/runs/1456020785)
* Tcl [2598](https://github.com/chipsalliance/Surelog/actions/runs/1446532862) vs. Py [78](https://github.com/chipsalliance/Surelog/actions/runs/1446532863)
* Tcl [2594](https://github.com/chipsalliance/Surelog/actions/runs/1441188538) vs. Py [74](https://github.com/chipsalliance/Surelog/actions/runs/1441188539)

### CI Builds

Below table lists the difference in build times for individual platform i.e. for each platform, the cell value is 'build time using Tcl' - 'build time using Python'. All times are in seconds.
```
+---------------+-------------+-------------+-------------+-------------+-------------+
|               | #2614 - #94 | #2610 - #90 | #2606 - #86 | #2598 - #78 | #2594 - #74 |
+---------------+-------------+-------------+-------------+-------------+-------------+
| Full Build    |       -2696 |        2554 |       -2923 |       -2348 |         982 | <= This tends to go either way since certain builds could have a long wait time
+---------------+-------------+-------------+-------------+-------------+-------------+
| coverage      |       -1904 |       -2047 |       -2136 |       -1986 |       -1636 | <= Coverage builds are failing on all Tcl runs. Ignored for bottom computation
| linux-gcc     |        -419 |         786 |        -276 |         579 |        -527 |
| macos-gcc     |        -540 |        -672 |        -362 |        -721 |        -739 |
| macos-clang   |         160 |         164 |        -283 |       -1718 |       -1569 |
| msys2-gcc     |        1440 |        3360 |        2340 |        2940 |        3000 |
| valgrind      |        -949 |       -1549 |       -1759 |       -1947 |       -1218 |
| windows-cl    |        -821 |        -540 |       -1245 |        -330 |        -240 |
+---------------+-------------+-------------+-------------+-------------+-------------+
| Largest Delta |        1440 |        3360 |        2220 |        2940 |        3000 | <= Longest running platform using Tcl - Longest running platform using Python
+---------------+-------------+-------------+-------------+-------------+-------------+
| Percentage    |       21.43 |       36.36 |       32.46 |       36.30 |       34.97 | <= This is the improvement in overall build time assuming all platforms kick off immediately
+---------------+-------------+-------------+-------------+-------------+-------------+
```
A few important deductions
* Msys2 builds are the slowest and modularized implementation improves the build times by over 30%.
* Ironically, windows-cl and macos-gcc builds go the opposite direction. Both are about 10% slower.
* Overall though, the CI builds finish ~20% faster since Msys2 happens to be the bottleneck.

### CI Regression

Below table lists the difference in regression run times for each platform i.e. regression run-time using Tcl - regression run-time using Python. All times are in seconds.
```
+-------------+-------------+-------------+-------------+-------------+-------------+
|             | #2594 - #74 | #2598 - #78 | #2606 - #86 | #2610 - #90 | #2614 - #94 |
+-------------+-------------+-------------+-------------+-------------+-------------+
| linux-gcc   |          59 |        -181 |        -282 |           1 |          34 |
| macos-clang |         557 |         512 |         266 |         315 |         393 |
| macos-gcc   |         180 |         226 |         136 |          49 |         175 |
| msys2-gcc   |        3218 |        2854 |        2352 |        3634 |        2028 |
| windows-cl  |         706 |         504 |         148 |         592 |         388 |
+-------------+-------------+-------------+-------------+-------------+-------------+
```
A few important deductions
* Once again, the biggest benefit is on Msys2 platform, followed by windows.
* Linux occasionally reports marginally slower runs using Python

### CI Regression Results

Below table lists the CPU and Memory comparison across the implementations. Note, only tests that run for at least 5 seconds and report at least 100MB memory usage are included. The variance in smaller tests is too high. All values are percentage difference with respect to the value reported by corresponding Tcl run. Also, note that these values are for linux-gcc platform. Unfortunately, none of the other platforms generate this similar statistics using Tcl.
```
+-----------------------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|                             |  (#2594 - #74)%   |  (#2598 - #78)%   |  (#2606 - #86)%   |  (#2610 - #90)%   |  (#2614 - #94)%   |
+                             +-------------------+-------------------+-------------------+-------------------+-------------------+
|                             |    Time |  Memory |    Time |  Memory |    Time |  Memory |    Time |  Memory |    Time |  Memory |
+-----------------------------+-------------------+-------------------+-------------------+-------------------+-------------------+
| AmiqEth                     |   16.67 |   -0.58 |   -5.26 |   -0.10 |  -28.57 |   -0.49 |    5.56 |    0.29 |    4.76 |    0.39 |
| AmiqSimpleTestSuite         |   16.67 |   -0.48 |  -16.67 |    0.00 |  -20.00 |   -0.48 |    0.00 |   -0.48 |    0.00 |   -0.96 |
| Ariane                      |   12.96 |    0.27 |  -14.55 |    0.41 |  -25.00 |    0.41 |    3.57 |    0.34 |    7.46 |    0.41 |
| ArianeElab2                 | ------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- |  -16.67 |   -2.38 |
| AzadiRTL                    |    8.89 |   -0.28 |  -14.89 |   -0.28 |  -34.29 |    0.19 |    2.17 |   -0.38 |    7.14 |    0.09 |
| BlackParrotConf             | ------- | ------- |  -16.67 |   -0.63 | ------- | ------- |    0.00 |    0.00 |    0.00 |   -2.48 |
| BuildOVMPkg                 |    3.70 |   -0.28 |  -19.35 |    0.14 |  -40.91 |   -0.14 |    0.00 |   -0.57 |    8.57 |   -0.28 |
| BuildUVMPkg                 |   10.88 |    0.44 |   -8.00 |   -0.06 |  -13.85 |    0.06 |   -4.23 |   -0.06 |    7.63 |   -0.13 |
| Compl1001                   |    5.83 |   -0.24 |  -11.72 |   -0.32 |  -19.63 |    0.00 |    0.81 |   -0.08 |    0.71 |   -0.12 |
| CoresSweRV                  |   20.59 |    0.64 |   -2.86 |    0.77 |  -14.29 |    0.58 |    6.25 |    0.71 |    7.69 |    0.84 |
| Earlgrey                    |   16.13 |    0.07 |   -9.38 |   -0.14 |  -30.43 |    0.07 |    0.00 |   -0.07 |    8.57 |    0.07 |
| Earlgrey_0_1                |   13.95 |   -0.10 |  -19.57 |   -0.21 |  -22.22 |   -0.05 |    2.38 |   -0.15 |    8.16 |    0.21 |
| Earlgrey_Verilator_01_05_21 |   39.39 |    6.26 |   29.17 |    5.88 |    9.62 |    6.59 |   33.33 |    6.30 |   36.71 |    6.22 |
| Earlgrey_Verilator_0_1      |   11.36 |   -0.30 |    2.08 |   -0.20 |  -18.42 |   -0.50 |    2.17 |   -0.64 |   11.54 |   -0.35 |
| Google                      |   14.42 |    1.90 |    2.56 |    2.28 |  -25.00 |    2.58 |    4.42 |    2.39 |   12.70 |    2.49 |
| Ibex                        |   13.33 |   -0.29 |   -3.03 |   -0.10 |  -20.00 |   -0.48 |    3.23 |    0.00 |   11.43 |   -0.19 |
| IbexGoogle                  |   13.64 |    0.00 |   -4.00 |   -0.55 |  -27.78 |   -0.74 |    8.33 |    0.18 |   11.11 |   -0.73 |
| Icarus                      |    8.33 |   -1.30 |    0.00 |   -1.32 |  -33.33 |   -2.60 |    7.69 |    1.32 |   20.00 |   -3.85 |
| IncompTitan                 |   15.38 |   -0.40 |   -1.82 |   -0.51 |  -29.27 |   -0.30 |    7.27 |   -0.10 |   13.11 |    0.00 |
| MiniAmiq                    |    0.00 |   -0.66 |  -12.50 |    0.00 | ------- | ------- |   12.50 |   -0.67 |   11.11 |    0.33 |
| Monitor                     |   12.50 |    0.00 |    0.00 |   -1.25 |  -14.29 |   -0.31 |   11.11 |   -1.25 |   20.00 |   -0.63 |
| NyuziProcessor              |   13.33 |   -0.30 |   -6.25 |    0.00 |  -33.33 |   -0.74 |    6.25 |   -0.44 |   11.11 |   -0.15 |
| OVMSwitch                   |   12.50 |   -1.20 |  -12.50 |   -1.59 |  -50.00 |   -0.79 |   11.11 |   -1.18 |   10.00 |   -1.19 |
| Opentitan                   |   12.82 |   -0.06 |    0.00 |   -0.43 |  -21.88 |   -0.37 |   13.95 |   -0.06 |   15.22 |   -0.31 |
| Rsd                         |   23.53 |   -0.21 |   11.43 |   -0.24 |   -1.75 |   -0.21 |   22.37 |   -0.34 |   24.36 |   -0.05 |
| Scr1                        |    0.00 |   -0.51 |    0.00 |   -0.25 |  -20.00 |   -0.51 |   20.00 |   -1.01 |   13.33 |   -0.25 |
| SimpleVMM                   |    7.69 |    0.00 |   -7.14 |   -0.84 |  -27.27 |   -0.84 |   12.50 |   -0.42 |    0.00 |   -1.27 |
| UVMNestedSeq                | ------- | ------- | ------- | ------- | ------- | ------- |   16.67 |   -2.11 | ------- | ------- |
| UVMSwitch                   |   10.00 |   -1.97 |  -20.00 |   -1.66 |  -37.50 |   -2.30 |   16.67 |   -2.30 |    0.00 |   -2.30 |
| Unisim                      |   15.13 |    0.00 |   -4.88 |    0.35 |  -12.15 |    1.34 |   15.49 |    0.44 |   15.28 |    1.43 |
| UtdSV                       |   25.00 |   -0.57 |   -6.67 |   -1.14 |    0.00 |   -0.57 |   26.32 |    0.00 |   21.05 |    1.74 |
| Verilator                   |   13.33 |   -0.17 |  -11.36 |   -0.69 |  -27.78 |   -0.17 |   15.09 |   -0.17 |    1.92 |    0.17 |
| Yosys                       |   10.00 |    0.00 |  -10.00 |   -1.26 |  -37.50 |    0.64 |   23.08 |   -1.90 |    0.00 |    0.65 |
| YosysBigSimAmber23          | ------- | ------- | ------- | ------- | ------- | ------- |   16.67 |   -1.45 |    0.00 |    1.47 |
| YosysBigSimLm32             |   14.29 |    0.00 |  -33.33 |   -0.83 |  -16.67 |    0.41 |    0.00 |    0.83 |    0.00 |    0.41 |
| YosysBigSimOpenMsp          |   16.67 |   -1.87 |    0.00 |   -0.95 | ------- | ------- |    0.00 |   -2.36 |    0.00 |   -0.96 |
| YosysIce40                  |   15.38 |    0.52 |   -7.69 |   -0.78 |  -40.00 |   -0.78 |    7.14 |   -0.52 |    0.00 |   -0.52 |
| YosysOldOr                  |    9.09 |   -0.24 |    0.00 |   -0.12 |  -10.00 |   -0.48 |   15.38 |   -0.24 |    7.69 |   -0.24 |
| YosysOpenSparc              |   14.93 |    6.07 |    2.63 |    5.31 |  -16.67 |    5.19 |   18.75 |    6.11 |   14.46 |    5.00 |
| YosysSmallBoom              |   29.73 |    0.46 |    5.56 |    0.38 |   -3.23 |    1.41 |   25.00 |    0.23 |   27.27 |    3.01 |
| YosysTestSuite              |   18.18 |   -0.64 |    0.00 |   -1.92 |  -22.22 |   -1.92 |    9.09 |    1.31 |    8.33 |   -1.30 |
| YosysTests                  |   15.38 |   -0.12 |   -1.85 |   -0.25 |  -29.27 |    0.13 |   13.79 |   -0.13 |   11.67 |    0.00 |
+-----------------------------+-------------------+-------------------+-------------------+-------------------+-------------------+
```
A few important deductions
* Memory reporting using Python is usually around 2% away in either direction from one reported by Tcl
* CPU time variance is rather high and once again can go in either direction

In summary -
* Linux builds using Python are marginally slower
* Overall build times across all platforms improved by over 30%
* Both implementations are generating _similar_ regression outputs
* Variance in CPU reporting for regression tests is rather high

If all this looks good, we should be ready to toggle the switch to use Python whenever the authoritative powers wish.

**Comment by alaindargelas** [comment]
OK, please submit the PR. I'll approve it.
